agent/otel/otlpmqttexporter/otlp.go

```diff
@@ -0,0 +1,118 @@
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/pmetric/pmetricotlp"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+type exporter struct {
+	// Input configuration.
+	config     *Config
+	client     *http.Client
+	tracesURL  string
+	metricsURL string
+	logsURL    string
+	logger     *zap.Logger
+	settings   component.TelemetrySettings
+	// Default user-agent header.
+	userAgent string
+}
+
+// Crete new exporter.
+func newExporter(cfg config.Exporter, set component.ExporterCreateSettings) (*exporter, error) {
+	oCfg := cfg.(*Config)
+
+	if oCfg.Address != "" {
+		_, err := url.Parse(oCfg.Address)
+		if err != nil {
+			return nil, errors.New("address must be a valid mqtt server")
+		}
+	}
+
+	userAgent := fmt.Sprintf("%s/%s (%s/%s)",
+		set.BuildInfo.Description, set.BuildInfo.Version, runtime.GOOS, runtime.GOARCH)
+
+	// Client construction is deferred to start
+	return &exporter{
+		config:    oCfg,
+		logger:    set.Logger,
+		userAgent: userAgent,
+		settings:  set.TelemetrySettings,
+	}, nil
+}
+
+// start actually creates the HTTP client. The client construction is deferred till this point as this
+// is the only place we get hold of Extensions which are required to construct auth round tripper.
+func (e *exporter) start(_ context.Context, _ component.Host) error {
+	token := e.config.Client
+	if token == nil {
+		opts := mqtt.NewClientOptions().AddBroker(e.config.Address).SetClientID(e.config.Id)
+		opts.SetUsername(e.config.Id)
+		opts.SetPassword(e.config.Key)
+		opts.SetKeepAlive(10 * time.Second)
+		opts.SetDefaultPublishHandler(func(client mqtt.Client, message mqtt.Message) {
+			e.logger.Info("message on unknown channel, ignoring", zap.String("topic", message.Topic()), zap.ByteString("payload", message.Payload()))
+		})
+		opts.SetPingTimeout(5 * time.Second)
+		opts.SetAutoReconnect(true)
+
+		if e.config.TLS {
+			opts.TLSConfig = &tls.Config{InsecureSkipVerify: true}
+		}
+
+		client := mqtt.NewClient(opts)
+		if token := client.Connect(); token.Wait() && token.Error() != nil {
+			return token.Error()
+		}
+		e.config.Client = client
+	}
+
+	return nil
+}
+
+func (e *exporter) pushTraces(_ context.Context, _ ptrace.Traces) error {
+	return fmt.Errorf("not implemented")
+}
+
+// pushMetrics Exports metrics
+func (e *exporter) pushMetrics(ctx context.Context, md pmetric.Metrics) error {
+	tr := pmetricotlp.NewRequest()
+	tr.SetMetrics(md)
+	request, err := tr.MarshalJSON()
+	if err != nil {
+		return consumererror.NewPermanent(err)
+	}
+	return e.export(ctx, e.config.MetricsTopic, request)
+}
+
+func (e *exporter) pushLogs(_ context.Context, _ plog.Logs) error {
+	return fmt.Errorf("not implemented")
+}
+
+func (e *exporter) export(_ context.Context, metricsTopic string, request []byte) error {
+	// convert metrics to interface
+	body := request
+	if token := e.config.Client.Publish(metricsTopic, 1, false, body); token.Wait() && token.Error() != nil {
```

Comment:
so I think we need to embed the body in an RPC call, the way we do here
https://github.com/ns1labs/orb/blob/9fef3976bed6c47bb2f733c7931838395f79cb2e/agent/backend/pktvisor/pktvisor.go#L345

Reply:
hmm this maybe is the reason sinker is failing.